pilot_experiment/functions/calculations.py

```diff
@@ -0,0 +1,245 @@
+'''
+Author: Alexandra Lee
+Date Created: 17 April 2020
+
+Scripts to process expression data for pilot experiment
+'''
+import pandas as pd
+import os
+import random
+import numpy as np
+import pickle
+
+import warnings
+
+def fxn():
+    warnings.warn("deprecated", DeprecationWarning)
+
+with warnings.catch_warnings():
+    warnings.simplefilter("ignore")
+    fxn()
+    
+np.random.seed(123)
+
+def get_coexpression_stats(correlation_file,
+operon_annotation_file,
+core_gene_ids_file,
+acc_gene_ids_file,
+coexpression_threshold):
+    """
+    Returns summary statistics about co-expressed and co-operonic genes
+
+    Arguments
+    ----------
+    correlation_file: str
+        Pickle file containing pairwise correlation between all genes
+
+    operon_annotation_file: str
+        File containing list of genes per operon
+
+    core_gene_id_file: str
+        File containing core genes annotated from annotate_genes() in process_data.py
+
+    acc_gene_id_file: str
+        File containing accessory genes annotated from annotate_genes() in process_data.py
+
+    coexpression_threshold: float
+        Hard threshold cutoff that determine if genes are co-expressed .
+        If corr(gene i, gene j) > coexpression_threshold, then gene i and j are co-expressed 
+    """
+    # Read correlation data
+    core_gene_ids = pickle.load(open(core_gene_ids_file, "rb"))
+    acc_gene_ids = pickle.load(open(acc_gene_ids_file, "rb"))
+    corr_data = pickle.load(open(correlation_file, "rb"))
+
+    # Get all gene ids
+    all_gene_ids = list(corr_data.index)
+
+    # Define threshold for highly co-expressed genes
+    threshold = coexpression_threshold
+
+    # Apply threshold to identify which genes are co-expressed
+    coexpressed = corr_data>threshold
+
+    # Get upper triangle of correlation matrix
+    coexpressed_triu = pd.DataFrame(data=np.triu(coexpressed,1),
+                                            index=coexpressed.index,
+                                            columns=coexpressed.columns)
+
+    # Get total number of genes that are co-expressed per gene
+    # This will serve as a test case
+    num_coexpressed_genes = coexpressed_triu.sum(axis=1)
```

Comment:
I really appreciate you double checking. So maybe I am missing something. My logic was the following:
1. I start with my correlation network `corr_data` which is a gene x gene matrix that is symmetric
2. Then I apply my threshold cutoff `coexpressed = corr_data>threshold` which returns a boolean matrix that is still symmetric
3. Since the matrix is symmetric, then if (geneA, gene B) = 1 then (geneB,geneA) =1. Both of these are telling me that gene A is correlated with gene B based on my threshold cutoff. And so I only want to count 1 of them as opposed to 2. This is why I took the upper triangle. But more precisely, +1 off the upper triangle, because I also wanted to discard relationship between genes and itself.

Please let me know if my logic doesn't make sense to you and if I should update my calculations function. Definitely something that should be addressed sooner rather than later

Reply:
I see what you mean, but I still think that if you're trying to calculate the number of genes each gene is correlated with (or equivalently, the degree of each gene in the binarized network), you do actually want to double-count each edge.

As an example, take this super simple network: ![graph](https://user-images.githubusercontent.com/2345877/81946296-321fca80-95cd-11ea-85b2-eb62a7a0a2a3.png)

Each node is connected to 2 other nodes. It has this adjacency matrix:
```
[[0, 1, 1],
 [1, 0, 1],
 [1, 1, 0]]
```
So taking the sum of the rows or columns gives the correct degree (2) for each node. But if you take the upper triangle:
```
[[0, 1, 1],
 [0, 0, 1],
 [0, 0, 0]]
```
then taking the sum of rows or columns undercounts the degree of 2 of the nodes.

Does that make sense? If this example is too simplistic let me know, there might be something I'm missing about what you're trying to do.

Reply:
AH you're correct! Thank you so much for being a very thorough reviewer and finding this bug! I was accidentally in thinking about these matrices in the same way as the correlation matrices so I will go ahead and fix this